flink-streaming-java/src/test/java/org/apache/flink/streaming/api/operators/sink/WriterOperatorTest.java

```diff
@@ -0,0 +1,345 @@
+
+		// remove CheckpointBarrier which is not expected
+		testHarness.getOutput()
+			.removeIf(record -> record instanceof CheckpointBarrier);
+
+		Assert.assertTrue(verifyPreCommitOutput.test(testHarness.getOutput().toArray()));
+
+		testHarness.endInput();
+		testHarness.waitForTaskCompletion();
+		Assert.assertTrue(verifyEndOutput.test(testHarness.getOutput().toArray()));
+
+		AbstractWriterOperator<Integer, Tuple3<Integer, Long, Long>> s =  testHarness.getHeadOperator();
+
+		DummyWriter writer = (DummyWriter) s.getWriter();
+		Assert.assertTrue(writer.isClosed());
+
+		// set the operator state from previous attempt into the restored one
+		return taskStateManagerMock.getLastJobManagerTaskStateSnapshot();
+	}
+
+	static final class StatelessWriterSink implements TestSink<Integer, Tuple3<Integer, Long, Long>, Tuple3<Integer, Long, Long>, Void> {
+
+		@Override
+		public Writer<Integer, Tuple3<Integer, Long, Long>, Tuple3<Integer, Long, Long>> createWriter(
+			InitContext context, List<Tuple3<Integer, Long, Long>> states) {
+			return new DummyWriter();
+		}
+	}
+
+	static final class StatefulWriterSink implements TestSink<Integer, Tuple3<Integer, Long, Long>, Tuple3<Integer, Long, Long>, Void> {
+
+		@Override
+		public Writer<Integer, Tuple3<Integer, Long, Long>, Tuple3<Integer, Long, Long>> createWriter(InitContext context, List<Tuple3<Integer, Long, Long>> states) {
+			return new DummyWriter(3, states);
+		}
+
+		@Override
+		public Optional<SimpleVersionedSerializer<Tuple3<Integer, Long, Long>>> getWriterStateSerializer() {
+			return Optional.of(new WriterStateSerializer());
+		}
+	}
+
+	static final class DummyWriter
+		implements Writer<Integer, Tuple3<Integer, Long, Long>, Tuple3<Integer, Long, Long>> {
+
+		static final Tuple3<Integer, Long, Long> LAST_ELEMENT = Tuple3.of(Integer.MAX_VALUE, Long.MAX_VALUE, Long.MAX_VALUE);
+
+		private boolean isClosed;
+
+		private final int maxCacheElementNum;
+
+		// element, timestamp, watermark
+		private List<Tuple3<Integer, Long, Long>> elements;
+
+		DummyWriter(int maxCacheElementNum, List<Tuple3<Integer, Long, Long>> restoreElements) {
+			this.isClosed = false;
+			this.elements = new ArrayList<>(restoreElements);
+			this.maxCacheElementNum = maxCacheElementNum;
+		}
+
+		DummyWriter() {
+			this(0, Collections.emptyList());
+		}
+
+		@Override
+		public void write(Integer element, Context context) {
+			elements.add(Tuple3.of(element, context.timestamp(), context.currentWatermark()));
+		}
+
+		@Override
+		public List<Tuple3<Integer, Long, Long>> prepareCommit(boolean flush) {
+			final List<Tuple3<Integer, Long, Long>> r = elements;
+			if (flush) {
+				elements.add(LAST_ELEMENT);
+				return elements;
+			} else if (elements.size() >= maxCacheElementNum) {
+				elements = new ArrayList<>();
+				return r;
+			} else {
+				return Collections.emptyList();
+			}
+		}
+
+		@Override
+		public List<Tuple3<Integer, Long, Long>> snapshotState() {
+			return elements;
```

Comment:
Logically here we need to swap to a new list.

Reply:
This is by design. The `List` only be flush when the number is larger than `maxCacheElementNum`.